oeps/oep-0051-bp-conventional-commits.rst

```diff
@@ -0,0 +1,184 @@
+
+We are adopting the `Conventional Commits`_ spec, with our own adjustments.
+
+.. note::
+   Many people are familiar with the `Angular commit message format`_, which uses conventional commits and inspired the Conventional Commits spec.  We are **not** adopting the Angular rules.
+
+Git commit messages have these parts::
+
+    <type>: <subject>
+
+    <body>
+
+    <footer>
+
+Type
+----
+
+We use these commit types:
+
+* **build**: anything to do with building or releasing the repo: Makefile, tox.ini, Travis, Jenkins, GitHub actions, and so on.
+
+* **chore**: a repetitive mechanical task such as updating requirements or translations.
+
+* **docs**: documentation-only changes. Documentation changes for other types of work should  be included in those commits. This includes more than the formal docs for a repo, it also covers READMEs, ADRs, docstrings, and so on.
+
+* **feat**: a new feature, or a change to an existing feature. Anything that changes the set of features.  Public API entry points and their behavior are part of the feature set of a product, so changes to those should be marked with "feat".
+
+* **fix**: bug fixes, including changes to the behavior of features, and changes that remove or mitigate security vulnerabilities.
+
+* **perf**: performance improvement.
+
+* **refactor**: code reorganization that doesn't change behavior from the repo consumer perspective.
+
+* **revert**: undo a previous change. The previous commit message, including the type prefix, is included.
+
+* **style**: improve the styling of the code.
+
+* **test**: test-only changes. Adds missing tests or corrects existing tests. Tests accompanying other types of changes go into those commits.
+
+* **exp**: experimental or exploratory changes that are not meant to be permanent.  Occasionally changes have to be pushed to GitHub or even production that are intended to be short-lived. For example, debugging continuous integration, or ad-hoc debugging on live systems.
+
+
+Breaking changes include an exclamation point as part of the type::
+
+    feat!: remove the ability to author courses
+
+Ideally, a single commit won't mix these types together.  If a commit does, use the most important type in the commit message.  The priority order for the types is generally: revert, feat, fix, perf, docs, test, build, refactor, style, chore, exp.
+
+Scope
+-----
+
+The `Conventional Commits`_ spec includes a parenthesized scope after the type.  Open edX repos are large and varied, making standardization of scopes difficult.  We won't use scopes for now.  This could change in the future.
+
+Subject
+-------
+
+Commit message subjects should be short enough to fit on one line.  We aren't putting a hard limit on character length, but 70 characters is a good time to turn your attention to the body of the commit message.  Put more information in the body of the commit message to fully explain your change.  Jira or GitHub issue numbers can be included at the end of the subject.
```

Comment:
Please only include Jira issue numbers if the Jira issue is public.

Reply:
I think we should always include jira numbers if there is one, even if it isn’t public, but:
a) the commit message still needs to stand on its own; the ticket reference should be purely supplemental even if it is public
b) we should endeavor to keep as many of our jira projects and tickets open to the community as possible

Reply:
I definitely want the ticket numbers to be visible with `git log —oneline`, _especially_ with long log bodies; I think they should be put at the end of the subject but not “counted against” the length.

Reply:
@nedbat maybe we need to clarify that ticket/issue numbers _should_ be in the subject so they can be seen with `log —oneline`

Reply:
@inventhouse Can you explain why you believe it is important to have ticket numbers visible in the git log?

Reply:
@nasthagiri Wow, there's a lot to unpack; here are a few quick thoughts...

Tickets link to the past: As long as we use tickets of some kind to track work, they will continue to be significant historical artifacts for the changes that result; they will tend to have details and clues that were rightly (at the time) summarized away or simply missed in the commit log.  We must not lose them.

Tickets link to the future: Once a commit is merged to the main branch it can never be modified, but a referenced ticket can be updated with information or links after-the-fact; we have no other way to do this, and we must not lose that.

The connection from a commit to the significant ticket or tickets can be direct and reliable if the commit includes the reference, or (in my experience) a tenuous, unreliable, mini-sleuth (or not-so-mini!)

Even if the ticket is not currently public, it may be made public in the future, but the reference can't be added post-facto.  Even if it is not made public, the reference itself leaves a clue that can be inquired about; the easier it is to see if there might be more information, the easier it is for internal engineers to help the community.

I sincerely believe the importance of including ticket references, no matter how good the commit message might be, cannot be overstated.[^1]

Separately, most of the time I find `git log` much too verbose for showing more than a commit or two; the bigger picture is completely swamped by the details and I have no sense of "where I am" after a couple screens; thus I use `--oneline` a great deal.  Ticket references are the best, shortest, way we have to enable "popping open" more information right from the space-constrained subject line.  Although the full commit message may (hopefully) be better curated, it takes more steps to open and tends to lose my place in the list of logs; if the ticket looks promising, I'll certainly read the log too.  I don't want to lose this option, and I think putting the reference at the end of the subject is a good compromise.

[^1]: Seriously, I think ***every*** commit should reference a ticket, though I'm not (presently) proposing that requirement.

Reply:
(BTW: our OSPR tools looks in PR titles, not commit messages)

Reply:
I'm going to side with the "links in the body" side here.  The subject is best used for words that explain.  Supporting information, including references to Jira or GitHub issues should be in the body.  It's harder to get to, but we're asking people to write longer commit messages. We need to get better at using those messages.

Reply:
Oh well; I made my case and I’ve started to get some ideas for tools I can write to help work without the easy ticket references; hopefully they can help others as well.

Reply:
(But _definitely_ include those references in the commit message!)